### ISSUE TITLE
Respect semver 2.0.0

### ISSUE THREAD

**Issue by SladeThe** [labstack/echo-jwt#12]
Greetings!
According to [go.dev](https://go.dev/ref/mod), every `go.mod` dependency must respect [semantic versioning](https://semver.org/spec/v2.0.0.html), because many go tools rely on it.
A lib maintainer can NOT ignore the rule just because he wishes to.
I strongly suggest to make `echo-jwt` follow all common Golang agreements.
Otherwise, I would recommend everyone to avoid this lib and use their own or 3rd-party JWT middleware.

**Comment by aldas** [comment]
https://github.com/labstack/echo-jwt/pull/9#issuecomment-1649588437

@SladeThe there is nothing that requires you use this middleware.  50+% of it is different configuration options. 

If you absolutely disagree and can not use this lib then try this

This is stripped down version of same middleware:
```go
func jwtMW(signingKey []byte) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			_, auth, ok := strings.Cut(c.Request().Header.Get("Authorization"), "bearer ")
			if !ok {
				return echo.NewHTTPError(http.StatusUnauthorized, "authorization header missing or invalid")
			}

			token, err := jwt.ParseWithClaims(auth, jwt.MapClaims{}, func(token *jwt.Token) (interface{}, error) {
				if token.Method.Alg() != "HS256" {
					return nil, echo.NewHTTPError(http.StatusUnauthorized, "unexpected jwt signing method")
				}
				return signingKey, nil
			})
			if err != nil {
				return err
			}
			if !token.Valid {
				return echo.NewHTTPError(http.StatusUnauthorized, "invalid token")
			}
			c.Set("token", token)
			return next(c)
		}
	}
}
```



Full example:
```go
package main

import (
	"errors"
	"fmt"
	"github.com/golang-jwt/jwt/v5"
	"github.com/labstack/echo/v4"
	"log"
	"net/http"
	"strings"
)

func jwtMW(signingKey []byte) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			_, auth, ok := strings.Cut(c.Request().Header.Get("Authorization"), "bearer ")
			if !ok {
				return echo.NewHTTPError(http.StatusUnauthorized, "authorization header missing or invalid")
			}

			token, err := jwt.ParseWithClaims(auth, jwt.MapClaims{}, func(token *jwt.Token) (interface{}, error) {
				if token.Method.Alg() != "HS256" {
					return nil, echo.NewHTTPError(http.StatusUnauthorized, "unexpected jwt signing method")
				}
				return signingKey, nil
			})
			if err != nil {
				return err
			}
			if !token.Valid {
				return echo.NewHTTPError(http.StatusUnauthorized, "invalid token")
			}
			c.Set("token", token)
			return next(c)
		}
	}
}

func main() {
	e := echo.New()

	e.Use(jwtMW([]byte("secret")))

	e.GET("/test", func(c echo.Context) error {
		token := c.Get("token")

		return c.JSON(http.StatusOK, fmt.Sprintf("token: %v", token))
	})

	if err := e.Start(":8080"); err != nil && !errors.Is(err, http.ErrServerClosed) {
		log.Fatal(err)
	}
}
```

Output:
```bash
x@x:~/code/echo$ curl -H "Authorization: bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ" http://localhost:8080/test
"token: \u0026{eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ 0xc0000120a8 map[alg:HS256 typ:JWT] map[admin:true name:John Doe sub:1234567890] [76 149 64 247 147 171 51 177 54 112 22 155 223 68 76 30 177 195 112 71 241 142 134 25 129 225 78 52 88 123 30 4] true}"
```

**Comment by aldas** [comment]
@vishr and @lammel it is up to you.

**Comment by SladeThe** [comment]
Currently, I use this middleware.
The last minor update broke my project in runtime, though compilation was successful.
I agree that integration tests find it out, but it happens AFTER the change is already made and deployed.
If it's going to happen again, I'm afraid I'll have to follow your advice and use copy-pasted code instead of `go.mod` dependency.

**Comment by SladeThe** [comment]
I said nothing about prod.
The app was deployed to dev, which didn't affect users, but other developers.

Right, following semver doesn't protect us from ALL bugs.
But it greatly reduces the number of bugs and make updates more predictable.

Imagine all libs ignore semver. It would be nearly impossible to build a project with many dependencies as each one requires time to maintain.

**Comment by lammel** [comment]
@SladeThe What is the actual problem that we are discussing here?

We have a 14 commit old library just 8 months old and only existing to workaround an unmaintained jwt lib and wrap it for abstraction. (Not providing any more patches is at least very compatible with semver :)

I guess you are referring to the README section about VERSIONING:

> This repository does not use semantic versioning. MAJOR version tracks which Echo version should be used. MINOR version tracks API changes (possibly backwards incompatible) and PATCH version is incremented for fixes.
> 
> For Echo v4 use v4.x.y releases. Minimal needed Echo versions
> - v4.0.0 needs Echo v4.7.0+
> main branch is compatible with the latest Echo version.

What is your suggestion for a versioning scheme to use for our use case considering the requirements above?

**Comment by aldas** [comment]
As this lib README states - "**This repository does not use semantic versioning**" since [#1](https://github.com/labstack/echo-jwt/pull/1) first pull request.  Which means that whatever major version should be treated as any other library with `0.X.Y` version.

What we can do better (until we decide do change versioning strategy):
* make that statement more visible (bold?) in README and clearer warning about it being 0.x.y.
* provide example for stripped down example that you can copy/paste if you do not accept it not using versioning strategy of your liking. I think current example can be even shorter. If I recall correctly goJWT had already some infrastructure for that.
* provide integration test (use middleware + handler through serveHTTP) example so you can catch these problems in your test/CI flow.

**Comment by SladeThe** [comment]
It looks like I'm talking to a brick wall.
I've removed the lib from my project.
The issue is no longer relevant.

**Comment by lammel** [comment]
> It looks like I'm talking to a brick wall. I've removed the lib from my project. The issue is no longer relevant.

As you did not come up with any constructive feedback or suggestion on how this could be handled better, I guess we are out of options here. 

What I can not understand is how you perceive our feedback as talking to a brick wall. But anyway, Good luck with your projects.

**Comment by aldas** [comment]
lets not get into flame war here.  

TLDR: so far versioning strategy for this library is not planned to change

-----

We are dealing conflicting world views here.

We have displeased middleware user that got burned by last release and is voicing his disappointment for us not using semantic versioning and therefore abusing Go package manager system with current release tags.
   
This is definitely valid concern. We knew this is a problem and did it anyway.  

What is being proposed is (probably): switch to semantic versioning, retract the latest release in `go.mod` and make new  release with incremented major version `v5.0.0`. 

Our responses indicate that we are not going to change this outlandish versioning strategy (brick wall). Comparing feedback to brick wall is valid - as so far we are not planning to change.

Everything is simple until we need to factor in that `golang-jwt` can have API changes in future (their pace has ramped up with new maintainers) and our plans to have `v5` for Echo and support `v4` in parallel for some time.  Now considering these 2 factors versioning is not anymore one dimensional linear progression - it is multidimensional.  And blame me for this decision - I decided - screw this - we are on perpetual `0.x.y` versioning strategy (FOR THIS REPO) except every Echo version has their own major number.  This is of course in conflict what  [go mod](https://go.dev/ref/mod#modules-overview) expects what Major version >2 means. README says what this library is promising and what not. We try to do things that we promise. It is up to user to accept these choices or not. Read before adding library to their project. 

Again - these are weak arguments or acceptable arguments - depending on the point of observation.  

SladeThe did the correct choice, as his beliefs expect how versioning should be done, and opted out using this middleware. This is perfectly fine as we live in the free world, use open source software. 

I'll add PR(s) for these 3 things mentioned in previous [post](https://github.com/labstack/echo-jwt/issues/12#issuecomment-1651097372) to ease people troubles with our disappointing choices.

**Comment by aldas** [comment]
For history sake. We removed JWT middleware from core in [v4.13.0](https://github.com/labstack/echo/releases/tag/v4.13.0) with breaking change and suggest this repo as alternative.  

I have created example for bare minimum JWT middleware that does not rely on this repository. This can be found here https://github.com/labstack/echo-jwt/discussions/29#discussion-7634738 If you do not need fancy features - copy pasting that example is probably enough.
